### ISSUE TITLE
Is there a way to obfuscate the zip URL action?

### ISSUE THREAD

**Issue by ghost** [aerni/statamic-zipper#7]
Hi there,

sorry if this is an obvious question but I did not see it in the docs...

Wondering if there is a way to hide the URL from the address bar when zipper does its action? Because it parses the /assets/... link into the URL for each file being zipped... so if the user accidentally refreshed the page or something it would just redownload the file, and it also allows them to see the URLs which is not optimal

I have tried adding statamic's obfuscate modifier to the zipper tags, to get a 500 error.

**Comment by aerni** [comment]
No, this is not possible. That's how actions work. I don't see the problem with this. It's just a link to a zip asset. That's no different to any other link, e.g. a link to an image.

And the link in the address bar shouldn't change in the first place. How are you using the addon?

**Comment by aerni** [comment]
But I might be able to obfuscate the link with a hash or something. But the way it works would stay the same.

**Comment by ghost** [comment]
The issue is, that if the user can see the link to the direct files being zipped, they can then share those links directly without sharing the files - and anyone could then download them without passing the purchase validation.

**Comment by ghost** [comment]
![Screen Shot 2022-05-30 at 1 41 01 PM](https://user-images.githubusercontent.com/75390655/171057017-61c95be6-23f7-4d0c-b31e-9f21961ed39c.png)


Here is what i mean. When i click a button which links to the {{ zip }} tag, it works fine, but it shows this in the address bar

**Comment by ghost** [comment]
```
  <a href="{{ zip:archive :filename='title' }}">
        <button>Download {{ title }}
        </button>
        </a>

**Comment by aerni** [comment]
So if I understand correctly, your real issue is that your files are behind a paywall and you don't want people being able to download the files without paying first? If yes, then obfuscating the links wouldn't solve your underlying issue. The link to the zip would still be shareable and work even if you don't pay for the product.

**Comment by ghost** [comment]
Yes, the music album (individual music songs being zipped into a zip file) can be previewed and then purchased. On successful purchase, download button shows up to download the album. To avoid needing a user to upload both a zip archive version and individual files (not good on server space) we need to have something which zips up the individual files.
Your plugin is the only one for statamic which does this.

So what you are saying is that, even if the zip file isn't on the zerver, if someone was to call the /!/obfuscatedurl that it would download the zip file or initiate the zipping?

Is there another way to implement this which avoids the URL parsing?

I still think /!/obfuscated is much better than /!/allthefiles.wav&allthefiles1.wave&...

If you could somehow obfuscate the links and have the obfuscation change every few hours, that would fix the underlying issue of sharing links

**Comment by aerni** [comment]
I agree that a short link might be better than listing all the URL's of the files to be zipped. I will consider this for a future update. 

However, your issue won't be solved by adding this feature. This addon can only zip assets that are public. If you want to prevent people from downloading the songs before they are purchased, I highly recommend you use a [private container](https://statamic.dev/assets/#private-containers). Otherwise anyone can download the songs if they know the URL.

Depending on your purchase flow you could also use [signed URL's](https://laravel.com/docs/9.x/urls#signed-urls) so a zip can only be downloaded by a specific user or only for a limited amount of time.

This addon won't solve your underlying problem. You will have to do some custom coding to achieve what you need. I suggest you look into [Laravel Zipstream](https://github.com/stechstudio/laravel-zipstream), which is the package that this addon uses to generate zips. It's pretty straightforward.

**Comment by aerni** [comment]
I actually might have come up with a solution that could work for you. I will keep you posted.

**Comment by aerni** [comment]
You can check out PR #8, which adds support for private containers and enhances security.

**Comment by ghost** [comment]
Wow, i really appreciate that you took the time to consider this feature and add support for it so quickly!

**Comment by aerni** [comment]
You're welcome. Your main concern remains though. Anyone with the link will be able to download your files. You'll have to add some sort of authentication yourself.

**Comment by aerni** [comment]
This will take more time than I thought. I came across some major issues when implementing the encrypted file path. You can follow along in #8 if you're interested.

**Comment by ghost** [comment]
> You're welcome. Your main concern remains though. Anyone with the link will be able to download your files. You'll have to add some sort of authentication yourself.

Thank you, i do in fact have a form of validation taking place using the CSRF token - and i believe if I serve the front end through something like react instead I can route URLs alot differently to the front end.

Thanks again and will follow along in #8 for the next updates

**Comment by aerni** [comment]
Just released v1.6.0. Note, that it requires Laravel 9.
